droidlet/interpreter/craftassist/mc_interpreter.py

```diff
@@ -496,63 +495,89 @@ def new_tasks():
+        d_receiver = d.get("receiver", None)
 
         tasks = []
-        for obj in objs:
-            item_stack = agent.get_item_stack(obj.eid)
-            idm = (item_stack.item.id, item_stack.item.meta)
-            task_data = {"eid": obj.eid, "idm": idm, "obj_memid": obj.memid}
-            tasks.append(self.task_objects["drop"](agent, task_data))
+        if ref_pickup is None and d_receiver is None:
+            raise ErrorWithResponse(
+                "I think you want me to get or put something, but I don't understand what"
+            )
 
-        return maybe_bundle_task_list(agent, tasks)
+        if ref_pickup:
+            objs = self.subinterpret["reference_objects"](self, speaker, ref_pickup)
+            objs = [obj for obj in objs if isinstance(obj, ItemStackNode)]
+            if not objs:
+                raise ErrorWithResponse(
+                    "I think you want me to get or put something, but I don't understand what"
+                )
+            # no selector, just pick the nearest object to agent from the list.  if agent is carrying one it should be nearest
+            # FIXME: do this properly, add a nearest to agent selector
+            if ref_pickup.get("filters", {}).get("selector") is None:
+                min_dist_obj = None
+                min_dist = 10000000
+                for obj in objs:
+                    dist = np.linalg.norm(np.array(obj.pos) - np.array(agent.pos))
+                    if dist <= min_dist:
+                        min_dist_obj = obj
+                        min_dist = dist
+                objs = [obj]
+
+            for obj in objs:
+                # some or all of the objects to deliver/drop may already be in inventory...
+                # only pick up objects already on the ground.
+                if "_on_ground" in obj.get_tags():
+                    task_data = {
+                        "pos": obj.pos,
+                        "eid": obj.eid,
+                        "obj_memid": obj.memid,
+                    }
+                    tasks.append(self.task_objects["get"](agent, task_data))
+
+        if d_receiver:
+            if len(objs) == 0:
+                raise ErrorWithResponse("I don't understand what you want me to get or place.")
+            try:
+                if d_receiver.get("reference_object"):
+                    # this is bring to an entity; but for now will just drop it at entities location
+                    mems = self.subinterpret["reference_locations"](self, speaker, d_receiver)
+                else:  # d_receiver.get("location"):
+                    assert d_receiver.get("location") is not None
+                    mems = self.subinterpret["reference_locations"](
+                        self, speaker, d_receiver["location"]
+                    )
+            except NextDialogueStep:
+                # TODO allow for clarification
+                raise ErrorWithResponse("I don't understand where you want me to move.")
```

Comment:
nit: more informative error response: "I don't understand where you want me to place the item."